File: src/main/java/com/kelsos/mbrc/controller/Controller.java
package com.kelsos.mbrc.controller;

import android.content.Intent;
import android.os.IBinder;
import android.util.Log;
import com.google.inject.Inject;
import com.google.inject.Injector;
import com.google.inject.Singleton;
import com.kelsos.mbrc.BuildConfig;
import com.kelsos.mbrc.configuration.Configuration;
import com.kelsos.mbrc.constants.UserInputEventType;
import com.kelsos.mbrc.events.MessageEvent;
import com.kelsos.mbrc.interfaces.ICommand;
import com.kelsos.mbrc.interfaces.IEvent;
import com.squareup.otto.Bus;
import com.squareup.otto.Subscribe;
import roboguice.service.RoboService;

import java.util.HashMap;
import java.util.Map;

@Singleton
public class Controller extends RoboService {

    @Inject private Injector injector;
    @Inject private Bus bus;
    private Map<String, Class<? extends ICommand>> commandMap;

    public Controller() {
        if (BuildConfig.DEBUG) {
            Log.d("mbrc-log", "Controller initialized");
        }
    }

    @Override public IBinder onBind(Intent intent) {
        return null;
    }

    public void register(String type, Class<? extends ICommand> command) {
        if (commandMap == null) {
            commandMap = new HashMap<String, Class<? extends ICommand>>();
        }
        if (!commandMap.containsKey(type)) {
            commandMap.put(type, command);
        }
    }

    public void unregister(String type, Class<? extends ICommand> command) {
        if (commandMap.containsKey(type) && commandMap.get(type).equals(command)) {
            commandMap.remove(type);
        }
    }

    /**
     * Takes a MessageEvent and passes it to the command execution function.
     *
     * @param event The message received.
     */
    @Subscribe public void handleUserActionEvents(MessageEvent event) {
        executeCommand(event);
    }

    public void executeCommand(IEvent event) {
        Class<? extends ICommand> commandClass = this.commandMap.get(event.getType());
        if (commandClass == null) return;
        ICommand commandInstance;
        try {
            commandInstance = injector.getInstance(commandClass);
            if (commandInstance == null) return;
            commandInstance.execute(event);
        } catch (Exception ex) {
            if (BuildConfig.DEBUG) {
                Log.d("mbrc-log", "executing command for type: \t" + event.getType(), ex);
                Log.d("mbrc-log", "command data: \t" + event.getData());
            }
        }

    }

    @Override public int onStartCommand(Intent intent, int flags, int startId) {
        Configuration.initialize(this);
        bus.register(this);
        return super.onStartCommand(intent, flags, startId);
    }

    @Override public void onStart(Intent intent, int startId) {
        super.onStart(intent, startId);
    }

    @Override public void onDestroy() {
        executeCommand(new MessageEvent(UserInputEventType.CancelNotification));
        super.onDestroy();
    }
}

File: app/src/main/kotlin/com/kelsos/mbrc/controller/Controller.kt
package com.kelsos.mbrc.controller

import android.app.Service
import android.content.Intent
import android.os.IBinder
import javax.inject.Inject
import javax.inject.Singleton
import com.kelsos.mbrc.events.ChangeWebSocketStatusEvent
import com.kelsos.mbrc.extensions.initDBFlow
import com.kelsos.mbrc.interactors.LibrarySyncInteractor
import com.kelsos.mbrc.messaging.NotificationService
import com.kelsos.mbrc.messaging.SocketMessageHandler
import com.kelsos.mbrc.net.SocketService
import com.kelsos.mbrc.receivers.PlayerActionReceiver
import com.kelsos.mbrc.receivers.StateBroadcastReceiver
import com.kelsos.mbrc.services.ServiceDiscovery
import com.kelsos.mbrc.utilities.RxBus
import com.kelsos.mbrc.utilities.SettingsManager
import com.raizlabs.android.dbflow.config.FlowManager
import roboguice.RoboGuice
import rx.Observable
import timber.log.Timber

@Singleton class Controller : Service() {

  @Inject private lateinit var socket: SocketService
  @Inject private lateinit var handler: SocketMessageHandler
  @Inject private lateinit var receiver: StateBroadcastReceiver
  @Inject private lateinit var actionReceiver: PlayerActionReceiver
  @Inject private lateinit var notificationService: NotificationService
  @Inject private lateinit var discovery: ServiceDiscovery
  @Inject private lateinit var settingsManager: SettingsManager
  @Inject private lateinit var bus: RxBus
  @Inject private lateinit var sync: LibrarySyncInteractor

  init {
    Timber.d("Application Controller Initialized")
  }

  override fun onBind(intent: Intent): IBinder? {
    return null
  }

  override fun onCreate() {
    super.onCreate()
    this.initDBFlow()
    RoboGuice.getInjector(this).injectMembers(this)
    this.registerReceiver(actionReceiver, actionReceiver.intentFilter)
    this.registerReceiver(receiver, receiver.intentFilter)
    bus.register(this,
        ChangeWebSocketStatusEvent::class.java,
        { this.onWebSocketActionRequest(it) })
  }


  override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
    Timber.v("[Service] start command received")
    Observable.merge(discovery.startDiscovery(), settingsManager.observableDefault)
        .first()
        .subscribe({
          if (it != null) {
            socket.startWebSocket()
            sync.sync()
          }

        }) { Timber.v(it, "Discovery failed") }

    return super.onStartCommand(intent, flags, startId)
  }

  override fun onDestroy() {
    super.onDestroy()
    Timber.v("[Service] destroying service")
    bus.unregister(this)
    notificationService.cancelNotification(NotificationService.NOW_PLAYING_PLACEHOLDER)
    socket.disconnect()
    FlowManager.destroy()
    this.unregisterReceiver(receiver)
    this.unregisterReceiver(actionReceiver)
  }

  private fun onWebSocketActionRequest(event: ChangeWebSocketStatusEvent) {
    when (event.action) {
      ChangeWebSocketStatusEvent.CONNECT -> {
        Timber.v("Attempting to start the websocket")
        socket.startWebSocket()
      }
      ChangeWebSocketStatusEvent.DISCONNECT -> {
        Timber.v("Attempting to stop the websocket")
        socket.disconnect()
      }
    }
  }

}
